File: ProjectEulerPython/math_helpers.py
import list_helpers
import math
from prime_factorization import PrimeFactorization, PrimeFactorizationEntry

def divisor_count(n: int, proper: bool = False, ordered_primes: [int] = None) -> int:
  """Returns the number of positive divisors of n.

  This uses the following formula:
     If n = p^a * q^b * r^c *... is a prime factorization of n, then the number of divisors, d(n) = (a + 1)(b + 1)(c + 1)...
     For proof, see: https://math.stackexchange.com/questions/433848/prime-factors-number-of-divisors

  Example:
    12 has proper positive divisors 1, 2, 3, 4, and 6; 12 is included if proper = False
  """
  result = 1
  factorization = prime_factorization(n, ordered_primes)
  for entry in factorization.entries():
    result *= entry.power + 1

  if proper:
    result -= 1

  return result

def divisors(n: int, proper: bool = False, ordered_primes: [int] = None) -> [int]:
  """Returns a list of all (proper) divisors of the integer n.

  Args:
    n: Divisors will be provided for this integer.
    proper: If False, all divisors will be provided. If True, only proper divisors will be provided (i.e. divisors not equal to n).
    ordered_primes: If provided, this should be a list of all prime numbers in increasing order between 1 and the square root of n (inclusive).

  """
  factorization = prime_factorization(n, ordered_primes)
  divisor_factorization = PrimeFactorization()
  for entry in factorization.entries():
    divisor_factorization.add_entry(entry.prime, 0)

  result = []
  product = 1
  while True:
    if not proper or product != n:
      result.append(product)

    incremented = False
    for divisor_entry in divisor_factorization.entries():
      if divisor_entry.power < factorization.entry(divisor_entry.prime).power:
        product *= divisor_entry.prime
        divisor_entry.power += 1
        incremented = True
        break
      else:
        for i in range(0, divisor_entry.power):
          product = product // divisor_entry.prime
        divisor_entry.power = 0
    if not incremented:
      break

  return result

def index_prime_by_sieve(n: int) -> [bool]:
  """Returns a boolean list where the ith entry (i.e. index) is True if i is a prime number."""
  result = [True for i in range(n + 1)]
  result[0] = False
  result[1] = False
  for i in range(2, int(math.sqrt(n)) + 1):
    if result[i]:
      for j in range(i * i, n + 1, i):
        result[j] = False
  return result

def lexicographic_permute(elements: [object]) -> bool:
  """Modifies the list (in-place) to the next lexicographical permutation. Returns True if the next permutation is the initial one, False otherwise.

  Examples:
    [A, C, B] would be modified to [B, A, C] and False would be returned
    [C, B, A] would be modified to [A, B, C] and True would be returned

  """
  last_index = len(elements) - 1

  if last_index < 1:
    return True

  i = last_index - 1
  while i >= 0 and not elements[i] < elements[i + 1]:
    i -= 1

  result = False
  if i < 0:
    result = True
    elements.reverse() # reverse all elements
  else:
    j = last_index
    while j > i + 1 and not elements[j] > elements[i]:
      j -= 1
    #swap elements[i] with elements[j]
    temp = elements[i]
    elements[i] = elements[j]
    elements[j] = temp
    list_helpers.reverse_elements(elements, i + 1) # reverse the elements (in place) indexed i + 1 through the end of the list

  return result


def nth_triangular_number(n: int) -> int:
  """Returns the nth triangular number; for proof see: https://en.wikipedia.org/wiki/Triangular_number"""
  return n * (n + 1) // 2

def prime_factorization(n: int, ordered_primes: [int] = None) -> PrimeFactorization:
  """Returns a prime factorization for the integer n.

  Args:
    n: The integer (target) for the prime factorization.
    ordered_primes: If provided it should be a comprehensive increasing list of all primes less than or equal to the square root of n.

  Example:
    360 = 2^3 * 3^2 * 5

  """
  result = PrimeFactorization()
  power = 0

  if (n < 0):
    n = -n

  if ordered_primes is None:
    # 2, the only even prime, is treated separately so the for loop below is more efficient (i.e. it can increment by 2 instead of 1)
    while n % 2 == 0:
      n /= 2
      power += 1

    if power > 0:
      result.add_entry(2, power)

    for i in range(3, int(math.sqrt(n)) + 1, 2):
      power = 0
      while n % i == 0:
        n = n // i
        power += 1
      if power > 0:
        result.add_entry(i, power)
  else:
    for p in ordered_primes:
      power = 0
      while n % p == 0:
        n = n // p
        power += 1
      if power > 0:
        result.add_entry(p, power)

  if n > 2:
    result.add_entry(n, 1)

  return result

def primes_below(n: int) -> [int]:
  """Returns a list of all prime numbers in increasing order between 2 and the integer n."""
  result = []
  prime_sieve = index_prime_by_sieve(n)
  for i in range(2, len(prime_sieve)):
    if prime_sieve[i]:
      result.append(i)
  return result

File: ProjectEuler/MathHelpers.cs
using System.Collections.Generic;

namespace ProjectEuler
{
    public static class MathHelpers
    {
        /// <summary>
        /// Constructs a boolean array where each positive entry > 1 will be true of the integer index is a composite number, false otherwise.
        /// </summary>
        /// <param name="n">Upper bound for the sieve.</param>
        public static bool[] IndexCompositeBySieve(long n)
        {
            var composite = new bool[n + 1];
            for (var i = 2L; i * i <= n; ++i)
            {
                if (!composite[i])
                {
                    for (var j = i * i; j <= n; j += i) composite[j] = true;
                }
            }
            return composite;
        }

        /// <summary>
        /// Constructs a list of all positive primes less than or equal to the integer provided.
        /// </summary>
        /// <param name="n">The upper bound for prime list.</param>
        public static List<long> PrimesLessOrEqualTo(long n)
        {
            var composite = IndexCompositeBySieve(n);

            var result = new List<long>();
            for (var i = 2L; i <= n; ++i)
            {
                if (!composite[i]) result.Add(i);
            }
            return result;
        }

        /// <summary>
        /// Constructs a prime factorization of the integer provided.
        /// <example>For example, 60 = 2^2 * 3 * 5.</example>
        /// </summary>
        /// <param name="n">The integer for which a prime factorization will be constructed.</param>
        public static PrimeFactorization PrimeFactorization(long n)
        {
            var result = new PrimeFactorization();
            var power = 0;

            while (n % 2L == 0)
            {
                n /= 2L;
                ++power;
            }
            if (power > 0) result.AddEntry(2L, power);

            for (var i = 3L; i <= System.Math.Sqrt(n); i += 2L)
            {
                power = 0;
                while (n % i == 0)
                {
                    n /= i;
                    ++power;
                }
                if (power > 0) result.AddEntry(i, power);
            }

            if (n > 2)
            {
                result.AddEntry(n, 1);
            }

            return result;
        }

        /// <summary>
        /// Calculates the sum of all positive primes less than or equal to the integer provided.
        /// <example>For example, 8 returns 17 = 2 + 3 + 5 + 7.</example>
        /// </summary>
        /// <param name="n">Upper bound for prime summation.</param>
        public static long SumPrimesLessOrEqualTo(long n)
        {
            var composite = IndexCompositeBySieve(n);
            var result = 0L;

            for (var i = 2L; i <= n; ++i)
            {
                if (!composite[i]) result += i;
            }
            return result;
        }

        /// <summary>
        /// Calcuates the nth triangular number (see https://en.wikipedia.org/wiki/Triangular_number).
        /// <example>For example, 6 returns 21.</example>
        /// </summary>
        /// <param name="n">The index of the triangular number to calculate.</param>
        public static long NthTriangularNumber(int n)
        {
            long i = n;
            return i * (i + 1) / 2;
        }

        /// <summary>
        /// Calculates the number of positive divisors of the integer provided.
        /// <example>For example, 18 returns 6 since it has divisors: 1, 2, 3, 6, 9, and 18.</example>
        /// </summary>
        /// <param name="n">The integer to count divisors for.</param>
        /// <returns>The total number of positive divisors of the integer n.</returns>
        public static int DivisorCount(long n)
        {
            var result = 1;
            var factorization = PrimeFactorization(n);
            foreach (var item in factorization.Entries)
            {
                result *= item.Power + 1;
            }
            return result;
        }

        /// <summary>
        /// Calculates all positive divisors of the integer provided.
        /// <example>For example, 18 returns IEnumerable<1, 2, 3, 6, 9, 18>.</example>
        /// </summary>
        /// <param name="n">The integer to calculate divisors for.</param>
        /// <param name="proper">If true, the integer n will be excluded.</param>
        /// <returns>All positive divisors of the integer n.</returns>
        public static IEnumerable<long> Divisors(long n, bool proper = false)
        {
            var factorization = PrimeFactorization(n);
            var divisorFactorization = new PrimeFactorization();
            foreach (var item in factorization.Entries)
            {
                divisorFactorization.AddEntry(item.Prime, 0);
            }

            var product = 1L;
            while (true)
            {
                if (!proper || product != n) yield return product;

                var incremented = false;
                foreach (var item in divisorFactorization.Entries)
                {
                    if (item.Power < factorization.Entry(item.Prime).Power)
                    {
                        product *= item.Prime;
                        ++item.Power;
                        incremented = true;
                        break;
                    }
                    else
                    {
                        for (var i = 0; i < item.Power; ++i)
                        {
                            product /= item.Prime;
                        }
                        item.Power = 0;
                    }
                }
                if (!incremented) break;
            }
        }

        /// <summary>
        /// Calculates the sum of the two integers provided in string form. Useful when the computation is outside the standard C# numeric data type range/bounds.
        /// </summary>
        /// <returns>The sum of integer1 and integer2.</returns>
        public static string AddIntegerStrings(string integer1, string integer2)
        {
            var result = "";

            //this algorithm assumes integer2's length is greater than or equal to integer1's length, so swap if necessary
            var lengthDiff = integer2.Length - integer1.Length;
            if (lengthDiff < 0)
            {
                var temp = integer1;
                integer1 = integer2;
                integer2 = temp;
                lengthDiff = -lengthDiff;
            }
            var maxLength = System.Math.Max(integer1.Length, integer2.Length);

            var carry = 0;
            for (var i = maxLength - 1; i >= 0; --i)
            {
                var sum = ((int)(integer2[i] - '0')) + carry;
                if (i >= lengthDiff)
                {
                    sum += ((int)(integer1[i - lengthDiff] - '0'));
                }
                result += (char)(sum % 10 + '0');
                carry = sum / 10;
            }
            if (carry > 0) result += (char)(carry + '0');

            return result.Reverse();
        }

        /// <summary>
        /// Calculates the product of the two integers provided in string form. Useful when the computation is outside the standard C# numeric data type range/bounds.
        /// </summary>
        /// <returns>The product of integer1 and integer2.</returns>
        public static string MultiplyIntegerStrings(string integer1, string integer2)
        {
            var negative = false;
            if (integer1[0] == '-')
            {
                negative = !negative;
                integer1 = integer1.Substring(1);
            }
            if (integer2[0] == '-')
            {
                negative = !negative;
                integer2 = integer2.Substring(1);
            }

            var digitsInReverse = new int[integer1.Length + integer2.Length];
            var index1 = 0;
            var index2 = 0;

            for (var i = integer1.Length - 1; i >= 0; --i)
            {
                var carry = 0;
                index2 = 0;
                int digit1 = integer1[i] - '0';

                for (var j = integer2.Length - 1; j >= 0; --j)
                {
                    int digit2 = integer2[j] - '0';
                    var sum = digit1 * digit2 + digitsInReverse[index1 + index2] + carry;
                    carry = sum / 10;
                    digitsInReverse[index1 + index2] = sum % 10;
                    ++index2;
                }

                if (carry > 0) digitsInReverse[index1 + index2] += carry;
                ++index1;
            }

            // find index of first non-zero digit from right to left (i.e. preceeding zeros due to reverse order)
            var index = digitsInReverse.Length - 1;
            while (index >= 0 && digitsInReverse[index] == 0) --index;
            if (index < 0) return "0";
            string result = "";
            if (negative) result += '-';
            while (index >= 0) result += (digitsInReverse[index--]);
            return result;
        }

        /// <summary>
        /// Calculates the Collatz function for the integer provided (see https://en.wikipedia.org/wiki/Collatz_conjecture).
        /// </summary>
        /// <param name="n">The input for the Collatz function.</param>
        public static long CollatzFunction(long n)
        {
            if (n % 2L == 0L) return n / 2L;
            return 3L * n + 1L;
        }

        /// <summary>
        /// Calculates the number of iterations (i.e. length) needed for the Collatz function to result in 1 starting at the provided integer.
        /// </summary>
        /// <param name="n">The initial input to the Collatz function.</param>
        /// <returns>The Collatz length starting at the integer n.</returns>
        public static int CollatzLength(long n)
        {
            var result = 1;
            while (n != 1L)
            {
                n = CollatzFunction(n);
                ++result;
            }
            return result;
        }

        /// <summary>
        /// Calculates the next lexicographic permutation of the array provided.
        /// </summary>
        /// <param name="elements">The array of elements to permute (in-place).</param>
        /// <returns>Returns true if the resulting permutation is the initial one, false otherwise.</returns>
        public static bool LexicographicPermute(char[] elements)
        {
            var lastIndex = elements.Length - 1;
            if (lastIndex < 1) return true;

            var i = lastIndex - 1;
            while (i >= 0 && !(elements[i] < elements[i + 1])) --i;

            var result = false;
            if (i < 0)
            {
                result = true;
                elements.Reverse(0, lastIndex);
            }
            else
            {
                var j = lastIndex;
                while (j > i + 1 && !(elements[j] > elements[i])) --j;
                CharHelpers.Swap(ref elements[i], ref elements[j]);
                elements.Reverse(i + 1, lastIndex);
            }
            return result;
        }

        /// <summary>
        /// Calculates the greatest common divisor of the integers a and b using the Euclidean algorithm.
        /// <example>For example, if a = 15 and b = 24 then their gcd is 3.</example>
        /// </summary>
        /// <returns>The greatest common divisor of a and b.</returns>
        public static long Gcd(long a, long b)
        {
            while (b != 0L)
            {
                var t = b;
                b = a % b;
                a = t;
            }
            return a;
        }

        /// <summary>
        /// Calculates the least common multiple of a and b by computing their product and dividing by their greatest common divisor.
        /// <example>For example, if a = 15 and b = 24 then their lcm is 120.</example>
        /// </summary>
        /// <returns>The least common multiple of a and b.</returns>
        public static long Lcm(long a, long b)
        {
            var d = Gcd(a, b);
            return a * b / d;
        }
    }
}
